### DISCUSSION TITLE
Save/Apply button display and functioning problems

### DISCUSSION THREAD

**Discussion by cpaschen** [joomengine/Joomla-Component-Builder#873] [Q&A]
I created an admin for a component well over a year ago and need to make adjustments, but having problem solving a basic issue.

On the admin view (in front-end), the only buttons that appear are "Save & Close" and "Close".

I need to have a "Save" (or "Apply") button to the form, as well as a Save & New and Save as Copy.

I can't figure out WHY these buttons are not appearing in the toolbar.

I've looked into the view's view.html.php file (on the front-end), but the code in the addToolBar() function appears to be identical to another form that I have that IS showing all the toolbar buttons (other than the obvious difference for the name of the view in each of the code areas).

The only difference that I can see (other than different fields and some postsave code), is that the form that does not display the save/apply toolbar button has  a couple liked views on the form.
Also, the problem form, at the end of the view.html.php >> addToolBar() function has:

`$this->toolbar = JToolbar::getInstance();`

which is not in the other view's file.

I was able to manually get the apply button to work by adding:

`JToolBarHelper::apply('book.apply', 'JTOOLBAR_APPLY');`

within the first if clause ( ... ::checkString($this->referral))  section.

The apply button then appeared, but just functioned the same as a SAVE & CLOSE button, always returning the user to the previous screen (closing the form).

I'm not at all sure why this might be happening, unless it is related to the linked views on this form.

If anyone has ANY ideas where to look next, please let me know as I've been trying to track this down for a couple days.
And I'm not sure if this is somehow a JCB issue or just a Joomla code issue.

**Comment by cpaschen** [comment]
FYI ... I've found that I had some permission settings (Admin View >> Settings >> Permissions) that I thought might be causing the issue, but I have remove them and still no Apply/Save button.

I also removed all Linked Views and still no change (although I didn't expect Linked views to cause any change - just trying almost anything right now).

BTW ... here's what a 'working' admin view's view.html.php file's addtoolbar function looks like:

	```
protected function addToolBar()
	{

		// set help url for this view if found
		$help_url = Kr_booksHelper::getHelpUrl('pen_name');
		if (Kr_booksHelper::checkString($help_url))
		{
			JToolbarHelper::help('COM_KR_BOOKS_HELP_MANAGER', false, $help_url);
		}
		// now initiate the toolbar
		$this->toolbar = JToolbar::getInstance();
	}

	```

Here is what the one that is NOT working looks like:

```
protected function addToolBar()
	{JFactory::getApplication()->input->set('hidemainmenu', true);
		$user = JFactory::getUser();
		$userId	= $user->id;
		$isNew = $this->item->id == 0;

		JToolbarHelper::title( JText::_($isNew ? 'COM_KR_BOOKS_BOOK_NEW' : 'COM_KR_BOOKS_BOOK_EDIT'), 'pencil-2 article-add');
		// Built the actions for new and existing records.
		if (Kr_booksHelper::checkString($this->referral))
		{
			if ($this->canDo->get('core.create') && $isNew)
			{
				// We can create the record.
				JToolBarHelper::save('book.save', 'JTOOLBAR_SAVE');
			}
			elseif ($this->canDo->get('core.edit'))
			{
				// We can save the record.
				JToolBarHelper::save('book.save', 'JTOOLBAR_SAVE');
			}
			if ($isNew)
			{
				// Do not creat but cancel.
				JToolBarHelper::cancel('book.cancel', 'JTOOLBAR_CANCEL');
			}
			else
			{
				// We can close it.
				JToolBarHelper::cancel('book.cancel', 'JTOOLBAR_CLOSE');
			}
		}
		else
		{
			if ($isNew)
			{
				// For new records, check the create permission.
				if ($this->canDo->get('core.create'))
				{
					JToolBarHelper::apply('book.apply', 'JTOOLBAR_APPLY');
					JToolBarHelper::save('book.save', 'JTOOLBAR_SAVE');
					JToolBarHelper::custom('book.save2new', 'save-new.png', 'save-new_f2.png', 'JTOOLBAR_SAVE_AND_NEW', false);
				};
				JToolBarHelper::cancel('book.cancel', 'JTOOLBAR_CANCEL');
			}
			else
			{
				if ($this->canDo->get('core.edit'))
				{
					// We can save the new record
					JToolBarHelper::apply('book.apply', 'JTOOLBAR_APPLY');
					JToolBarHelper::save('book.save', 'JTOOLBAR_SAVE');
					// We can save this record, but check the create permission to see
					// if we can return to make a new one.
					if ($this->canDo->get('core.create'))
					{
						JToolBarHelper::custom('book.save2new', 'save-new.png', 'save-new_f2.png', 'JTOOLBAR_SAVE_AND_NEW', false);
					}
				}
				$canVersion = ($this->canDo->get('core.version') && $this->canDo->get('book.version'));
				if ($this->state->params->get('save_history', 1) && $this->canDo->get('core.edit') && $canVersion)
				{
					JToolbarHelper::versions('com_kr_books.book', $this->item->id);
				}
				if ($this->canDo->get('core.create'))
				{
					JToolBarHelper::custom('book.save2copy', 'save-copy.png', 'save-copy_f2.png', 'JTOOLBAR_SAVE_AS_COPY', false);
				}
				JToolBarHelper::cancel('book.cancel', 'JTOOLBAR_CLOSE');
			}
		}
		JToolbarHelper::divider();
		// set help url for this view if found
		$help_url = Kr_booksHelper::getHelpUrl('book');
		if (Kr_booksHelper::checkString($help_url))
		{
			JToolbarHelper::help('COM_KR_BOOKS_HELP_MANAGER', false, $help_url);
		}
		// now initiate the toolbar
		$this->toolbar = JToolbar::getInstance();
	}
```


I'm at quite a loss as to why JCB is creating such a small here for the first instance and why the second is so involved.

I think I've disabled almost all of the settings specific to this non-working page, and I've removed what I could.

I also tried to replicate the smaller, working code in place of the longer, non-working code, but that resulted in the toolbar not appearing at all.


Any explanation as to why JCB is building 2 completely different toolbar functions would be appreciated.
And a solution to getting the apply/save button to appear on initial create or edit of a record would be a huge help.

**Reply by Llewellynvdm** [reply]
Okay so this is not really an issue as much as it is a function to help manage the workflow of an application. I understand that there can be cases that don't want this kind of behavior and then it very easy to change it in general.

Yet explaining what and why this is happening the way it does take a little more understanding of the expected result, so let me give that first. The expected result is that you open an item to edit it "not to look" but edit or create it, and when your done you go back to either the looking at it area or the list of those items. To have this we need a way to remember the way back the these referral views, and again referral meaning the "looking at the data, or the list of items of the data". So when you come from a referral and you do a `save as copy` or  just save and keep the item open, **_we lose track of where you came from_**, and can't take you back to this referral area. So the way to get around this was to only show two buttons, which will force you back the way you came to not allow a unexpected behavior. Remember the variation of relationships are very large... in between different complex applications.

Okay that explained here is the solution out of this behavior... remove the referral information from the url. So to demonstrate this take the url that you are using to open any of these edit views, and remove the following: `&return=aHR0cHM....` and if this exist `&ref={string}` and `&refid={number}`. You will now see that all buttons will show. but when you save and close or just close the redirection may not work as expected. There is away to manage that... but it will require some custom coding to be added.

Well this is my first quick response to your question... I need to run... but have a look and let me know if this already helps....

**Reply by cpaschen** [reply]
Well, not to be 'gushy', but whenever you provide explanations like this (or new youtube videos), I not only learn how to do things in JCB, but also learn a little more about Joomla itself. So THANKS for that explanation.  I definitely understand the approach.

And the good news is, YES, removing the &return solves the issue. Buttons show and the Save/Apply button works as it should.

But, of course, we need to have the Save & Close return to the previous/sending page (a list view of all the items).

Right now (without the &return value) it is returning to default site view for the component, which is not what we can have for this specific edit view.

If you could explain what needs to happen, I'll go wrangle the code and eventually try to post a step-by-step of what worked here later, in case others need to deal with this issue also.

**Comment by cpaschen** [comment]
BTW ... "it will require some custom coding to be added" ... do you mean custom code to JCB or just to the component?

If it is in the component, if you can give me a rough explanation I should be able figure that out. (Or point me to Joomla docs that might explain the problem/solution).

**Reply by Llewellynvdm** [reply]
The area in the current J3 controller that manages the [redirects are found here](https://github.com/Llewellynvdm/Joomla-Demo-Component/blob/master/site/controllers/look.php#L150). You should also look at these areas, the [save method](https://github.com/Llewellynvdm/Joomla-Demo-Component/blob/master/site/controllers/look.php#L260), and [cancel method](https://github.com/Llewellynvdm/Joomla-Demo-Component/blob/master/site/controllers/look.php#L203) as each of those also effect the redirection. Last area to look at it in the [view itself](https://github.com/Llewellynvdm/Joomla-Demo-Component/blob/master/site/views/look/view.html.php#L46), as this is where we catch these referral keys and act on it. You will see that the `$this->referral` value is [used around](https://github.com/Llewellynvdm/Joomla-Demo-Component/blob/master/site/views/look/view.html.php#L96) the view mini controller to trip buttons on and off and basically give the behavior I explained above.

To change any behavior you need to add/override with custom code ([here is how to add custom code](https://www.youtube.com/watch?v=DFMfIl-VkSk&list=PLQRGFI8XZ_wtGvPQZWBfDzzlERLQgpMRE&index=31)) to these areas, it will be tricky as you need a way to remember the area you want to go to once a successful close, save and close... etc.

_[see playlist:](https://www.youtube.com/playlist?list=PLQRGFI8XZ_wtGvPQZWBfDzzlERLQgpMRE)_
![image](https://user-images.githubusercontent.com/5607939/157976039-608c81b3-6bfc-4d27-9549-8ab2c2b84bb9.png)
